chapters/04_self-attention.tex

```diff
@@ -15,29 +15,63 @@ \section{Counting}
 
 \section{Table Lookup}
 
-Table lookup is just the special case of self-attention where the query and the key are equal.
+Given that position $i$ stores some integer $a_i$, a table lookup is a self-attention layer in which each position $i$ attends to position $a_j$. 
+If $f \colon [n] \to \R$ is any function and we can use the position embedding to store $f(j)$ at position $j$, then table lookup makes it possible to compute $f(a_i)$ at each position $i$.
 
+\iffalse
 \subsection{Single query}
 
-If there is a single query that is broadcast to all positions,
-there is a minimum distance $\delta$ between all keys, and we can compute $\delta$ at all positions,
-then just use a FFNN to test for equality.
-
-\citep{chiang-cholak-2022-parity}
+If every position has the same query ($q^{(i)} = q$ for all $i$) and there is a minimum distance $\delta$ between all keys ($|k^{(i)}-k^{(j)}| \ge \delta$ for all $i,j$), then we can just use a FFNN to compute $\mathbb{I}[k^{(i)} = q^{(i)}] \delta$, which is piecewise linear \citep{chiang-cholak-2022-parity}:
+\begin{center}
+  \begin{tikzpicture}
+    \begin{axis}[xtick={-1,0,1},xticklabels={$-\delta$,$0$,$\delta$,ytick={0,1},yticklabels={$0$,$\delta$}}}]
+      \addplot[mark=none] coordinates { (-5,0) (-1,0) (0,1) (1,0) (5,0) };
+    \end{axis}
+  \end{tikzpicture}
+\end{center}
+(Since this doesn't use attention, it doesn't strictly belong in this chapter.)
+\fi
 
 \subsection{Layernorm hash}
 
-\citep{merrill-sabharwal-2024-cot}
+\Citet{merrill-sabharwal-2024-cot} store integer $a_i$ as
+\begin{align*}
+  \textnormal{lh}(k) &= \textnormal{LayerNorm}\left(\begin{bmatrix} a_i \\ 1 \\ -a_i \\ -1 \end{bmatrix}\right) \\
+  &= \frac1{\sqrt{2a_i^2+2}} \begin{bmatrix} a_i \\ 1 \\ -a_i \\ -1 \end{bmatrix}.
+\end{align*}
+One advantage of this representation is that if we're only able to compute $a_i/i$ and $1/i$, we can still compute $\textnormal{lh}(a_i)$.
+
+So if we let
+\begin{align*}
+  q_i &= \textnormal{lh}(a_i) \\
+  k_j &= \textnormal{lh}(j)
+\end{align*}
+then the dot product $s_{i,j} = \textnormal{lh}(a_i) \cdot \textnormal{lh}(j)$ is uniquely maximized when $a_i = j$, so that self-attention will retrieve the value $v_j$ such that $a_i = j$.
+
+The other advantage of this representation is that $\textnormal{lh}(a) \cdot \textnormal{lh}(b)$ is uniquely maximized when $a=b$. 
+
+The disadvantage of this representation is that it requires the ability to selectively apply $\textnormal{LayerNorm}$ to just four components of a vector. We can't do this with standard transformer layers (either pre-norm or post-norm), but we can if we use pre-norm \emph{and} insert a linear transformation $\mathbf{W}^{(\textnormal{N})}$ before the layer normalization:
+\begin{align*}
+  \mathbf{y} &= \textnormal{Sublayer}(\textnormal{LayerNorm}(\mathbf{W}^{(\textnormal{N})} \mathbf{x})) + \mathbf{x}.
+\end{align*}
```

Comment:
```suggestion
\Citet{merrill-sabharwal-2024-cot} store integer $a_i$ as
\begin{align*}
  \textnormal{lh}(a_i) &= \textnormal{LayerNorm}\left(\begin{bmatrix} a_i \\ 1 \\ -a_i \\ -1 \end{bmatrix}\right) \\
  &= \frac1{\sqrt{2a_i^2+2}} \begin{bmatrix} a_i \\ 1 \\ -a_i \\ -1 \end{bmatrix}.
\end{align*}
One advantage of this representation is that if we're only able to compute $a_i/i$ and $1/i$ (as is common when counting positions using uniform attention), we can still compute $\textnormal{lh}(a_i)$. Another advantage is that $\textnormal{lh}(a)\cdot\textnormal{lh}(b)$ is uniquely maximized when $a=b$. That is, for any $i>0$, we have that $\textnormal{lh}(a_i)=\textnormal{LayerNorm}(a_i/i,1/i,-a_i/i,-1/i) = \textnormal{LayerNorm}(a_i,1,-a_i,-1)$. So if we let
\begin{align*}
  q_i &= \textnormal{lh}(a_i) \\
  k_j &= \textnormal{lh}(j)
\end{align*}
then the dot product $s_{i,j} = \textnormal{lh}(a_i) \cdot \textnormal{lh}(j)$ is uniquely maximized when $a_i = j$, so that self-attention will retrieve the value $v_j$ such that $a_i = j$. In the setting of hard-attention, this allows us to select only the position vector at position $j=a_i$, after which the the value of $f(j)$ can be retrieved by appropriately setting $W^{(V)}$.

The disadvantage of this representation is that it requires the ability to selectively apply $\textnormal{LayerNorm}$ to just four components of a vector. We can't do this with standard transformer layers (either pre-norm or post-norm), but we can if we use pre-norm \emph{and} insert a linear transformation $\mathbf{W}^{(\textnormal{N})}$ before the layer normalization:
\begin{align*}
  \mathbf{y} &= \textnormal{Sublayer}(\textnormal{LayerNorm}(\mathbf{W}^{(\textnormal{N})} \mathbf{x})) + \mathbf{x}.
\end{align*}
```

Reply:
Made some minor edits and added a little more detail